beetmoverscript/src/beetmoverscript/script.py

```diff
@@ -285,6 +290,208 @@ async def upload_data(context):
+
+     Sanity checking is done to ensure that no destination files would overwrite
+     each other.
+    """
+
+    concreteArtifactMap = []
+    errors = []
+
+    for map_ in artifactMap:
+        concretePaths = {}
+
+        full_glob_destinations = map_["paths"].get("*", {}).get("destinations")
+        other_paths = copy.deepcopy(map_["paths"])
+        if "*" in other_paths:
+            del other_paths["*"]
+
+        for taskId, artifacts in upstreamArtifactPaths.items():
+            if map_["taskId"] != taskId:
+                continue
+
+            for artifact in artifacts:
+                dest_artifact = remove_segments(artifact, strip_segments)
+
+                destinations = []
+                # We need to look at non-'*' paths separate from '*' paths.
+
+                for input_path, output in other_paths.items():
+                    # Skip any input paths that don't match the artifact name.
+                    if "*" in input_path:
+                        if not fnmatch.fnmatch(dest_artifact, input_path):
+                            continue
+                    else:
+                        if input_path != artifact:
+                            continue
+
+                    # If there's already destinations, we've already seen this artifact,
+                    # and we have a clash.
+                    if destinations:
+                        errors.append(f"'{artifact}' matched multiple concrete paths")
+                    else:
+                        destinations.extend(output["destinations"])
+
+                # If we didn't find any destinations for the artifact above, and
+                # there are full glob destinations, it should go there.
+                if full_glob_destinations and not destinations:
+                    destinations.extend(full_glob_destinations)
+
+                # If there are destinations, create a fully concrete entry,
+                # mapping the full path to the file on disk to destinations
+                # it belongs, with any requested segments removed.
+                if destinations:
+                    concretePaths[artifact] = {"destinations": []}
+                    for d in destinations:
+                        if not d.endswith(dest_artifact):
+                            d = os.path.normpath(f"{d}{dest_artifact}")
+                        concretePaths[artifact]["destinations"].append(d)
+
+        concreteArtifactMap.append(
+            {
+                "paths": concretePaths,
+                "taskId": map_["taskId"],
+            }
+        )
+
+    ensure_no_overwrites_in_artifact_map(concreteArtifactMap)
+
+    if errors:
+        raise ScriptWorkerTaskException(*errors)
+
+    return concreteArtifactMap
+
+
+def remove_segments(string, prefixes):
+    for p in prefixes:
+        if p in string:
+            string = string[string.find(p) + len(p) :]
+
+    return string
```

Comment:
This doesn't seem to do what it says on the tin, i.e. the "prefixes" aren't treated as prefixes, but as "if this shows up anywhere in `string`, drop it and anything before it".

Reply:
You're right. The original version of this did actually only deal in prefixes; I realized later that we had to deal with absolute paths, and ended up renaming most things. At the very least, I can adjust the argument's name. Ideally there would be a less ugly/hacky way to do this, but I haven't come up with one yet :(.

Reply:
I made a change which I think is an improvement: it now strips prefixes, but gains some smarts to always strip away the scriptworker-specific leading directories. Let me know what you think.